### ISSUE TITLE
E741 incorrectly flags variable as ambiguous

### ISSUE THREAD

**Issue by madman32** [PyCQA/pycodestyle#938]
The following example code is valid syntax (python 3.7/3.8 tested) but seems to cause the E741 ambiguous variable name 'l' failure (in the last line). It looks to me like that in this case L is perfectly valid and not ambiguous?

```
from __future__ import annotations

import pydantic

class Foo(pydantic.BaseModel):
    uid: str
    size: int

    @classmethod
    def from_dict(cls, creation_dict) -> Foo:
        uid = creation_dict["uid"]
        size = creation_dict["size"]
        return cls(uid=uid, size=size)


t_dict = {
    "t1" : [
        {
            "uid" : "blah1",
            "size": 0
        }
    ]
}
test = {
    l.uid : l for l in (Foo.from_dict(t) for t in t_dict["t1"])
}
```

**Comment by asottile** [comment]
See [this](https://www.flake8rules.com/rules/E741.html) for more details.  the latest pycodestyle version got better at identifying more places where assignments can occur (notably, in your case, loop variables)
